avocado/utils/vmimage.py

```diff
@@ -0,0 +1,375 @@
+        self.args = {'version': kwargs.get('version', self.default_version),
+                     'build': kwargs.get('build', self.default_build),
+                     'arch': kwargs.get('arch', self.default_arch),
+                     'checksum': kwargs.get('checksum', None),
+                     'algorithm': kwargs.get('algorithm', None),
+                     'cache_dir': kwargs.get('cache_dir',
+                                             tempfile.gettempdir())}
+
+
+class DebianImageProvider(ImageBase):
+
+    name = 'Debian'
+
+    def __init__(self, **kwargs):
+        super(DebianImageProvider, self).__init__(**kwargs)
+        # Here you can use string formatting tags corresponding to the
+        # arguments accepted (version, build and arch).
+        self.url_versions = 'https://cdimage.debian.org/cdimage/openstack/'
+        self.url_images = self.url_versions + '/{version}/'
+        self.image = 'debian-{version}-openstack-{arch}.qcow2$'
+
+        # Set the defaults to be used when the arguments are
+        # not provided. If you don't want to force a version,
+        # you can provide a regular expression that matches the
+        # corresponding format and we will get the highest
+        # occurrence of that expression.
+        self.default_version = '[0-9]+.[0-9]+.[0-9]+-.*'  # 9.1.4-20170830
+        self.default_build = None  # Build is part of the image version
+        self.default_arch = 'amd64'
+
+        # Dictionary of expected arguments and their values
+        self.args = {'version': kwargs.get('version', self.default_version),
+                     'build': kwargs.get('build', self.default_build),
+                     'arch': kwargs.get('arch', self.default_arch),
+                     'checksum': kwargs.get('checksum', None),
+                     'algorithm': kwargs.get('algorithm', None),
+                     'cache_dir': kwargs.get('cache_dir',
+                                             tempfile.gettempdir())}
+
+
+class ImageProviderProxy(object):
+    def __init__(self):
+        self.providers_list = []
+        self.providers_list.append(FedoraImageProvider)
+        self.providers_list.append(FedoraSecondaryImageProvider)
+        self.providers_list.append(CentOSImageProvider)
+        self.providers_list.append(UbuntuImageProvider)
+        self.providers_list.append(DebianImageProvider)
+
+    def get(self, name=None, **kwargs):
+        # No name, return the first registered provider
+        if name is None and self.providers_list:
+            return self.providers_list[0](**kwargs).get()
+        for provider in self.providers_list:
+            if provider.name.lower() == name.lower():
+                return provider(**kwargs).get()
+        raise AttributeError('No such provider (%s)' % name)
+
+    def list_providers(self):
+        return self.providers_list
+
+
+get = ImageProviderProxy().get
+list_providers = ImageProviderProxy().list_providers
```

Comment:
There is no need to create 2 instances just to get providers and function get...

Reply:
Yes, improving this.